### PR TITLE
resource/aws_transfer_server: Add support for VPC security groups

### LINKED ISSUE #15788: AWS Transfer Server: Attach VPC security groups at creation
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

You can now attach multiple security groups to a server in a VPC.

### New or Affected Resource(s)

* [`aws_transfer_server`](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/transfer_server)

### Potential Terraform Configuration

```hcl
resource "aws_transfer_server" "example" {
  endpoint_details {
    security_group_ids = [aws_security_group.example_1.id, aws_security_group.example_2.id]
  }
}
```

### References

[User Guide](https://docs.aws.amazon.com/transfer/latest/userguide/create-server-in-vpc.html).

Requires [AWS SDK v1.35.8](https://github.com/aws/aws-sdk-go/releases/tag/v1.35.8):

* #15500

Related:

* https://github.com/terraform-providers/terraform-provider-aws/pull/12599
* https://github.com/terraform-providers/terraform-provider-aws/pull/13737

[Note](https://docs.aws.amazon.com/transfer/latest/userguide/API_EndpointDetails.html#TransferFamily-Type-EndpointDetails-SecurityGroupIds):

> You can edit the SecurityGroupIds property in the UpdateServer API only if you are changing the EndpointType from PUBLIC or VPC_ENDPOINT to VPC. To change security groups associated with your server's VPC endpoint after creation, use the Amazon EC2 ModifyVpcEndpoint API.

### LINKED ISSUE #15823: aws_transfer_server - VPC Hosted never completes
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform CLI and Terraform AWS Provider Version
3.11.0

### Affected Resource(s)
aws_transfer_server

### Terraform Configuration Files

```
resource "aws_transfer_server" "sftp_transfer_server" {
  identity_provider_type = "SERVICE_MANAGED"
  logging_role           = data.aws_iam_role.ct_cloudwatch_role.arn
  endpoint_type = "VPC"

  endpoint_details {
      vpc_id = data.aws_vpc.selected.id
      subnet_ids = [data.aws_subnet.public_1c.id]
      address_allocation_ids = [data.aws_eip.transfer_server_eip.id]
  }

  tags = {
    NAME = var.server_name
    ENV  = var.env
  }
}
```

### Expected Behavior
transfer server should be created in a couple of minutes (UI confirmed)

### Actual Behavior
Resource creating waits indefinitely

```
aws_transfer_server.sftp_transfer_server: Still creating... [10m0s elapsed]
aws_transfer_server.sftp_transfer_server: Still creating... [10m10s elapsed]
aws_transfer_server.sftp_transfer_server: Still creating... [10m20s elapsed]
aws_transfer_server.sftp_transfer_server: Still creating... [10m30s elapsed]
...
```

### Steps to Reproduce

1. `terraform plan` - succeeds showing the expected resource
2. `terraform apply`

### Important Factoids

The VPC, Subnet, all networking (etc) are all shared from the master account, using AWS Control Tower.

* #0000

### LINKED ISSUE #16556: Transfer server cannot transition from public to an external VPC endpoint
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform CLI and Terraform AWS Provider Version

    Terraform v0.13.5
    provider[registry.terraform.io/hashicorp/aws] 3.13.*, 3.13.*, 3.13.*

### Affected Resource(s)

* `aws_transfer_server`

### Terraform Configuration Files

```hcl
resource "aws_transfer_server" "my_transfer_server" {
  identity_provider_type = "SERVICE_MANAGED"
+   endpoint_type          = "VPC"
+   endpoint_details {
+     subnet_ids             = xxx
+     vpc_id                 = xxx
+     address_allocation_ids = [xxx]
+   }
}
```

### Debug Output

Apologies as I do not have this available to report.

### Panic Output

Apologies again, not available to report.

### Expected Behavior

The transfer service should use the specified [external] VPC.

### Actual Behavior

Apply failed with _cannot add Address Allocation ID's at that time_.

### Steps to Reproduce

1. Create public endpoint AWS Transfer Family Instance.
2. Create VPC infrastructure for VPC type transfer family.
3. Change endpoint type to VPC, passing in the three necessary fields: VPC ID, Subnet IDs, Address Allocation IDs (see `+` lines above in the configuration file example)

### Important Factoids

Workaround: removing the Address Allocation ID's allows an apply to work, but creates an internal resource instead of external. Adding them back after this apply correctly creates the external type and associates the IP Address.

### LINKED ISSUE #19850: sweeper/aws_transfer_server: Cannot sweep transfer servers with API Gateway identity provider
The sweeper sets the `force_destroy` parameter before calling the resource Delete operation. When the identity provider is set to `API_GATEWAY`, this fails with the error

> InvalidRequestException: Cannot list users on server with IdP type: API_GATEWAY

Do not set `force_destroy` when identity provider is `API_GATEWAY`. Confirm if `force_destroy` can be set when the identity provider is `AWS_DIRECTORY_SERVICE`.

### Affected Resource

* aws_transfer_server

### PR DESCRIPTION
**Pull request by amadureira**
### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for pull request followers and do not help prioritize the request

Closes #15788.
Closes #19850.
Closes #16556.
Closes #15823.

Output from acceptance testing:

```
$ make testacc TESTARGS='-run=TestAccAWSTransferServer_basic'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -count 1 -parallel 20 -run=TestAccAWSTransferServer_basic -timeout 120m
=== RUN   TestAccAWSTransferServer_basic
=== PAUSE TestAccAWSTransferServer_basic
=== CONT  TestAccAWSTransferServer_basic
--- PASS: TestAccAWSTransferServer_basic (258.41s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	258.484s
$ make testacc TESTARGS='-run=testSweepTransferServers'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -count 1 -parallel 20 -run=testSweepTransferServers -timeout 120m
testing: warning: no tests to run
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	0.056s [no tests to run]
...
```

### CODE REVIEW

**Comment by amadureira** [comment]
Hi Folks, 
I found some problem in this PR. How can I put WIP label ?

**Comment by thiagonache** [comment]
> Hi Folks,
> I found some problem in this PR. How can I put WIP label ?

you can set in draft @amadureira

**Comment by thiagonache** [comment]
> > Hi Folks,
> > I found some problem in this PR. How can I put WIP label ?
> 
> you can set in draft @amadureira

Also, your PR should not change anything on the files CHANGELOG.md and hidden files.

**Comment by amadureira** [comment]
@teamterraform, 
I found this problem, but I think it's not a code problem:
https://github.com/hashicorp/terraform-provider-aws/pull/17539/checks?check_run_id=1915113884
Could you  ched

File: .changelog/17496.txt [hunk 1]
```diff
@@ -0,0 +1,3 @@
+```release-note:enhancement
+resource/aws_transfer_server: Add security_group_ids for  endpoint_details block
+```
```
Comment (ewbankkit) [inline .changelog/17496.txt:3]: Please delete this file as in applies to the now closed #17496 .

File: .changelog/17539.txt [hunk 2]
```diff
@@ -0,0 +1,3 @@
+```release-note:enhancement
+resource/aws_transfer_server: Add  security_group_ids parameter.
```
Comment (ewbankkit) [inline .changelog/17539.txt:2]:
```suggestion
resource/aws_transfer_server: Add `security_group_ids` attribute.
```

**Comment by ewbankkit** [comment]
Please add `Closes #15788` to the initial comment.

**Comment by pstrav** [comment]
Any updates on this?

**Comment by jaroslawksiazek** [comment]
Do we know, when this feature will be realised?

**Comment by amadureira** [comment]
Hi @ewbankkit,

Could you help-me again? I put the test result on first comment. I didn't found more trouble on this PR. Could you review it again? Let-me know if I forgot something.
thanks again.

@thiagonache 
I appreciate your review too!

**Comment by chiranjeevi-patel** [comment]
I have been waiting for this fix in terraform, as AWS sent a notice saying they are going to stop VPCENDPOINT support to aws transfer server by 31st Mar 2021, so we need to change the tranfer type to VPC asap, but the terraform module doesn't support VPC security group option

**Comment by thiagonache** [comment]
> Hi @ewbankkit,
> 
> Could you help-me again? I put the test result on first comment. I didn't found more trouble on this PR. Could you review it again? Let-me know if I forgot something.
> thanks again.
> 
> @thiagonache
> I appreciate your review too!

Hi @amadureira ,
It looks good to me but I'm not a contributor of this repo. 

We need @ewbankkit to give the final review and approve it.

**Comment by chiranjeevi-patel** [comment]
@ewbankkit  can you please review this PR

**Comment by sandeep3866** [comment]
any update on this feature ?

**Comment by chiranjeevi-patel** [comment]
Any updates on this PR @ewbankkit

**Comment by thiagonache** [comment]
It’s frustrating to see the community trying to help and being ignored.

**Comment by quinot** [comment]
@ewbankkit I've stumbled on this today, this feature would be immensely useful to me!

**Comment by johndotpub** [comment]
Ugh... we just ran into this issue too! Hopefully, this can be merged soon! 🤗

**Comment by NaveenKumarAvula** [comment]
We are really on the edge on this, Can this be merged please?

**Comment by puneetpunj** [comment]
My release is going live this month end, any idea when will this be merged?

**Comment by kapoorsumit2019** [comment]
Terraform team - Please let us know when the fix will be available. We're also using AWS Transfer for service in our project and as workaround we're using "null_resource" tio create "transfer-server" via terraform configuration.

**Review by ewbankkit — APPROVED** [review]
LGTM :rocket:.

#### Commercial

```console
% make testacc TEST=./aws TESTARGS='-run=TestAccAWSTransfer_serial/Server'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -count 1 -parallel 20 -run=TestAccAWSTransfer_serial/Server -timeout 180m
=== RUN   TestAccAWSTransfer_serial
=== RUN   TestAccAWSTransfer_serial/Server
=== RUN   TestAccAWSTransfer_serial/Server/disappears
=== RUN   TestAccAWSTransfer_serial/Server/Protocols
=== RUN   TestAccAWSTransfer_serial/Server/SecurityPolicy
=== RUN   TestAccAWSTransfer_serial/Server/UpdateEndpointTypePublicToVPC        
=== RUN   TestAccAWSTransfer_serial/Server/UpdateEndpointTypeVPCEndpointToVPCAddressAllocationIDs
=== RUN   TestAccAWSTransfer_serial/Server/VPCSecurityGroupIDs
=== RUN   TestAccAWSTransfer_serial/Server/basic
=== RUN   TestAccAWSTransfer_serial/Server/UpdateEndpointTypePublicToVPCAddressAllocationIDs
=== RUN   TestAccAWSTransfer_serial/Server/VPCAddressAllocationIDsSecurityGroupIDs
=== RUN   TestAccAWSTransfer_serial/Server/HostKey
=== RUN   TestAccAWSTransfer_serial/Server/Domain
=== RUN   TestAccAWSTransfer_serial/Server/UpdateEndpointTypeVPCEndpointToVPCSecurityGroupIDs
=== RUN   TestAccAWSTransfer_serial/Server/UpdateEndpointTypeVPCToPublic
=== RUN   TestAccAWSTransfer_serial/Server/APIGateway
=== RUN   TestAccAWSTransfer_serial/Server/ForceDestroy
=== RUN   TestAccAWSTransfer_serial/Server/UpdateEndpointTypeVPCEndpointToVPC
=== RUN   TestAccAWSTransfer_serial/Server/VPC
=== RUN   TestAccAWSTransfer_serial/Server/VPCAddressAllocationIDs
=== RUN   TestAccAWSTransfer_serial/Server/VPCEndpointID
=== RUN   TestAccAWSTransfer_serial/Server/APIGatewayForceDestroy
--- PASS: TestAccAWSTransfer_serial (3799.61s)
    --- PASS: TestAccAWSTransfer_serial/Server (3799.61s)
        --- PASS: TestAccAWSTransfer_serial/Server/disappears (151.55s)
        --- PASS: TestAccAWSTransfer_serial/Server/Protocols (283.05s)
        --- PASS: TestAccAWSTransfer_serial/Server/SecurityPolicy (207.53s)
        --- PASS: TestAccAWSTransfer_serial/Server/UpdateEndpointTypePublicToVPC (302.33s)
        --- PASS: TestAccAWSTransfer_serial/Server/UpdateEndpointTypeVPCEndpointToVPCAddressAllocationIDs (241.86s)
        --- PASS: TestAccAWSTransfer_serial/Server/VPCSecurityGroupIDs (75.81s)
        --- PASS: TestAccAWSTransfer_serial/Server/basic (211.54s)
        --- PASS: TestAccAWSTransfer_serial/Server/UpdateEndpointTypePublicToVPCAddressAllocationIDs (369.64s)
        --- PASS: TestAccAWSTransfer_serial/Server/VPCAddressAllocationIDsSecurityGroupIDs (206.68s)
        --- PASS: TestAccAWSTransfer_serial/Server/HostKey (184.54s)
        --- PASS: TestAccAWSTransfer_serial/Server/Domain (187.40s)
        --- PASS: TestAccAWSTransfer_serial/Server/UpdateEndpointTypeVPCEndpointToVPCSecurityGroupIDs (192.16s)
        --- PASS: TestAccAWSTransfer_serial/Server/UpdateEndpointTypeVPCToPublic (233.88s)
        --- PASS: TestAccAWSTransfer_serial/Server/APIGateway (196.55s)
        --- PASS: TestAccAWSTransfer_serial/Server/ForceDestroy (190.65s)
        --- PASS: TestAccAWSTransfer_serial/Server/UpdateEndpointTypeVPCEndpointToVPC (121.09s)
        --- PASS: TestAccAWSTransfer_serial/Server/VPC (184.91s)
        --- PASS: TestAccAWSTransfer_serial/Server/VPCAddressAllocationIDs (284.45s)
        --- PASS: TestAccAWSTransfer_serial/Server/VPCEndpointID (54.44s)
        --- PASS: TestAccAWSTransfer_serial/Server/APIGatewayForceDestroy (228.42s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	3803.238s
```

**Comment by ewbankkit** [comment]
@amadureira Thanks for the contribution :tada: :clap:.
I went ahead and pushed changes so that `SecurityGroupIds` can be modified after the Transfer Server has been created.

**Pull request merged**